src/main/java/org/spongepowered/common/world/SpongeTeleportHelper.java

```diff
@@ -47,205 +51,132 @@
 
-        // We've already checked zero right above this.
-        for (int currentLevel = 1; currentLevel <= height; currentLevel++) {
-            // Check above
-            safe = checkAroundLocation(location.add(0, currentLevel, 0), width);
-            if (safe.isPresent()) {
-                return safe;
-            }
+            // If the block isn't safe, no point in continuing on this run.
+            if (block.isSafeBody) {
 
-            // Check below
-            safe = checkAroundLocation(location.add(0, -currentLevel, 0), width);
-            if (safe.isPresent()) {
-                return safe;
-            }
-        }
+                // Check the block ABOVE is safe for the body, and the two BELOW are safe too.
+                if (getBlockData(currentTarget.add(0, 1, 0), world, blockCache).isSafeBody
+                        && isFloorSafe(currentTarget, world, blockCache)) {
 
-        return Optional.empty();
-    }
+                    // This position should be safe. Get the center of the block to spawn into.
+                    return true;
+                }
+            }
 
-    private Optional<Location<World>> checkAroundLocation(Location<World> location, final int radius) {
-        if (isSafeLocation(location.getExtent(), location.getBlockPosition())) {
-            return Optional.of(location);
-        }
+            return false;
+        }).findFirst();
 
-        // Now we're going to search in expanding concentric circles...
-        for (int currentRadius = 0; currentRadius <= radius; currentRadius++) {
-            Optional<Vector3i> safePosition = checkAroundSpecificDiameter(location, currentRadius);
-            if (safePosition.isPresent()) {
-                // If a safe area was found: Return the checkLoc, it is the safe
-                // location.
-                return Optional.of(new Location<>(location.getExtent(), safePosition.get()));
-            }
+        if (result.isPresent()) {
+            return Optional.of(new Location<>(world, result.get().toDouble().add(0.5, 0, 0.5)));
         }
 
+        // No vectors matched, so return an empty optional.
         return Optional.empty();
     }
 
-    private Optional<Vector3i> checkAroundSpecificDiameter(Location<World> checkLoc, final int radius) {
-        World world = checkLoc.getExtent();
-        Vector3i blockPos = checkLoc.getBlockPosition();
-        //example at radius 2
-        //..c..
-        //.....
-        //..o..
-        //.....
-        //.....
-        // Check out at the radius provided.
-        blockPos = blockPos.add(radius, 0, 0);
-        if (isSafeLocation(world, blockPos)) {
-            return Optional.of(blockPos);
-        }
-        //example at radius 2
-        //..c01
-        //.....
-        //..o..
-        //.....
-        //.....
-        // Move up to the first corner..
-        for (int i = 0; i < radius; i++) {
-            blockPos = blockPos.add(0, 0, i);
```

Comment:
You are right, I've clearly made an error when I touched this, I'm surprised it didn't get picked up sooner.
The previous behaviour before i touched it however was even more wrong, and unclear + uncommented.